anonlink/solving/_multiparty_solving_inner.cpp

```diff
@@ -0,0 +1,254 @@
+        size_t retval2 = this->set_or_increment(group1, group0, 1);
+        assert(retval1 == retval2);
+        return retval1;
+    }
+    
+    void merge_into(Group *absorber, Group *absorbee) {
+        assert(absorber);
+        assert(absorbee);
+        LinksStoreInnerInner &absorber_store = this->links_store_inner[absorber];
+        LinksStoreInnerInner &absorbee_store = this->links_store_inner[absorbee];
+        
+        // Merging two groups.
+        // They no longer need to keep track of common links.
+        absorber_store.erase(absorbee);
+        absorbee_store.erase(absorber);
+
+        // Move all the links from absorbee to absorber. Same with links referencing absorbee.
+        for (const auto &link_count : absorbee_store) {
+            Group *link = link_count.first;
+            assert (link);
+            size_t count = link_count.second;
+            assert (count);
+            
+            this->set_or_increment(absorber, link, count);
+            
+            LinksStoreInnerInner &mp_match = this->links_store_inner[link];
+            set_or_increment(mp_match, absorber, count);
+            mp_match.erase(absorbee);
+        }
+
+        // Cleanup.
+        this->links_store_inner.erase(absorbee);
+        if (absorber_store.empty()) {
+            this->links_store_inner.erase(absorber);
+        }
+    }
+    
+};
+
+
+void none_grouped(GroupsStore &groups_store, Record i0, Record i1) {
+    // Neither is in a group, so let's make one.
+    groups_store.make_group(i0, i1);
+}
+
+void one_grouped(GroupsStore &groups_store, LinksStore &links_store, Group *group, Record i) {
+    if (group->size() == 1) {
+        // We have two singletons (one has a group of itself, one doesn't), so we can merge them.
+        groups_store.add_to_group(group, i);
+    } else {
+        // The group has at least 2 elements. We've only matched with one so far (or else we'd
+        // already have a group of size at least one), so we can't merge.
+        Group *group_i = groups_store.make_group(i);
+        links_store.increment(group, group_i);
+    }
+}
+
+void two_grouped_merge(GroupsStore &groups_store,
+                       LinksStore &links_store,
+                       Group *absorber,
+                       Group *absorbee) {
+    groups_store.merge_into(absorber, absorbee);
+    links_store.merge_into(absorber, absorbee);
+}
+
+void two_grouped(GroupsStore &groups_store, LinksStore &links_store, Group *group0, Group *group1) {
+    size_t overlap = links_store.increment(group0, group1);
+    size_t group_0_size = group0->size();
+    size_t group_1_size = group1->size();
+    if (overlap == group_0_size * group_1_size) {
+        // Optimise by enlarging the bigger group.
+        if (group_0_size < group_1_size) {
+            two_grouped_merge(groups_store, links_store, group1, group0);
+        } else {
+            two_grouped_merge(groups_store, links_store, group0, group1);
+        }
+    }
+}
+
```

Comment:
All code above this point should be in an anonymous (aka unnamed) namespace to force internal linkage.

Reply:
Done!